dpp/src/data_contract/state_transition/data_contract_update_transition/validation/basic/schema_compatibility_validator.rs

```diff
@@ -0,0 +1,384 @@
+/// The Schema compatibility validator is a port of a JavaScript version
+/// https://bitbucket.org/atlassian/json-schema-diff-validator/src/master/
+///
+/// The functionality has been ported 'as is' without any logic improvements and optimizations
+use std::convert::TryFrom;
+
+use anyhow::Context;
+use itertools::Itertools;
+use json_patch::PatchOperation;
+use jsonptr::{Pointer, Resolve};
+use serde_json::Value as JsonValue;
+use thiserror::Error;
+
+const PROPERTY_REQUIRED: &str = "required";
+const PROPERTY_DEFINITIONS: &str = "definitions";
+const PROPERTY_REF: &str = "$ref";
+
+#[derive(Default, Debug, Clone)]
+pub struct ValidationOptions {
+    pub allow_new_one_of: bool,
+    pub allow_new_enum_value: bool,
+    pub allow_reorder: bool,
+    pub deprecated_items: Vec<String>,
+}
+
+struct RemovedItem {
+    name: String,
+    operation: PatchOperation,
+}
+
+#[derive(Error, Debug)]
+pub enum DiffVAlidatorError {
+    /// The error that happens when there there is a problem with access to the fields
+    /// in dynamic data structure
+    #[error("Error while validation: {0}")]
+    DataStructureError(#[from] anyhow::Error),
```

Comment:
Can't quite figure out where this macro is coming from, looks neat!

Reply:
the macro comes from `thiserror` crate. It forwards the implementation of `source` and `display` to the original error. In this case all not-business logic errors are `anyhow` errors. Anyhow stores the context, the `source` - the original error. This makes returned errors quite easy to understand.